_drafts/heartbeats.md

```diff
@@ -0,0 +1,963 @@
+robust solution for viewing logs in a timeline from a single device, I suggest
+generating a random (or better, monotonically increasing) integer on boot for
+use as a "boot ID" of sorts, and then use milliseconds-since-boot as the
+timestamp. It should work well enough given the device sends these logs to a
+server every so often
+
+### Resetting Data for Each Heartbeat
+
+Once the heartbeat interval has finished and the metrics have been flushed, each
+metric value should be reset.
+
+Let me try to emphasize how important this is with an example. Let's imagine my
+fictitious smart-toaster company is trying to measure how long, on average, all
+of our devices in the field are connected over Wi-Fi per hour. In an ideal
+world, every device is connected to Wi-Fi for all 3,600 seconds of each hour.
+
+This stat can either be recorded in two ways on the toaster firmware: as a
+continuous summation or as a heartbeat metric that is reset every hour.
+
+Below is a chart of a single device recording its time connected to Wi-Fi
+measured in both ways.
+
+<p align="center">
+  <img width="600" src="{% img_url device-metrics/wifi-time-connected.svg %}" />
+</p>
+
+When looking at this chart, it is clear by looking at the heartbeat metric (red
+line) in hour 3 that there was a dip in time connected. This is not immediately
+obvious when looking at a chart of a continuously summed number.
+
+To calculate the average time connected per hour, we use two different
+calculations.
+
+For the continuous sum, we can just divide the final number by the number of
+hours:
+
+```
+19800 seconds / 6 hours = 3,300 seconds per hour
+```
+
+or get this value by computing the deltas between each event received:
+
+```
+  (3600 - 0) + (7200 - 3600) + (9000 - 7200)
++ (12600 - 9000) + (16200 - 12600) + (19800 - 16200)
+= 19800 seconds
+
+19800 seconds / 6 hours = 3,300 seconds per hour
+```
+
+See how that got a little messy quickly?
+
+To compute this metric for the heartbeat data where the values are reset each
+hour, we just add all the intervals and divide by the number of heartbeats
+
+```
+(3600 + 3600 + 1800 + 3600 + 3600 + 3600) / 6 = 3,300 seconds per hour
+```
+
+The real issue comes into play when we want to compute a particular metric
+across all hours of a particular firmware version or group of devices. If every
+hour's metric is based upon the previous hour's data, for every single device,
+what happens if we miss a subset of heartbeats from a device?
+
+Do we extrapolate the data across the missing points? Do throw out the remainder
+of the device's session data until the device reboots? Maybe we should just
+ignore devices that have missing data in our queries.
+
+All of those are easily solved by just resetting the values for each heartbeat!
+If a heartbeat doesn't have a complete interval recorded, just throw it out. In
+the past, I've actually put this as a column in the database row,
+`is_full_hour`.
```

Comment:
May just drop this part ... I think the mention below in extra metadata captures the idea